analyzer/runtime/runtime.go

```diff
@@ -93,6 +97,12 @@ func (m *processor) PreWork(ctx context.Context) error {
 	return nil
 }
 
+// Implements block.BlockProcessor interface.
+func (m *processor) FinalizeFastSync(ctx context.Context, lastFastSyncHeight int64) error {
+	// For runtimes, fast sync does not disable any dead reckoning and does not ignore any updates.
```

Comment:
Yes, that's a good point. I've thought of those. As of this PR, the comment is true because dead reckoning is not disabled anywhere yet :). But my intention has actually been to keep it true (= keep dead reckoning even inside fast sync) longer-term too, unless we have strong incentives to disable it for individual values.

In general, for every value where we'll disable dead reckoning, I'm concerned about how to be reasonably sure that whatever is inside `FinalizeFastSync` brings us to the exact same state as dead reckoning would have. So my plan is to initially disable dead reckons only in a very limited set of cases:
- Consensus state that is easily fetchable from `StateToGenesis()`, so we can avoid a whole RPC during fast-sync. I think that's all or most of governance, scheduler, registry state.
- Native balances. The are also contained in consensus state. Because every transfer touches the fee accumulator, this will reduce db contention/deadlocks. More importantly, it will prevent us from generating negative balances, which the DB constraints disallow.
- Staking info in consensus. That, too, is full contained in `StateToGenesis()` so it's easy to implement `FinalizeFastSync()`. The reason to disable it is because this is the one value where dead-reckoning updates are non-commutative: Instead of just summing things, we're doing some fractional math. So if we process blocks out of order, the dead-reckoned values will be wrong anyway, so might as well go the extra mile and just not dead-reckon them in the first place.

Later we can add more, if any dead-reckoned value turns out to cause significant slowdowns. I expect the main problem will be db contention. But yes, when we do that, we'll have to be careful to have the right code in `FinalizeFastSync` to counteract the disabled dead reckons.

Reply:
Hmm i think that works for `evmTokens` since all the operations are commutative, but I think the `contracts.gasUsed` [insert](https://github.com/oasisprotocol/nexus/blob/main/analyzer/queries/queries.go#L379) needs to be updated to make it commutative as well, eg, if multiple `evm.Calls` are processed before the `evm.Create` is:
```
WITH
	contract_gas_used AS (
		SELECT SUM(gas_used) AS total_gas
		FROM chain.runtime_transactions
		WHERE runtime = $1 AND "to" = $2::text
	)
INSERT INTO chain.evm_contracts
    (runtime, contract_address, creation_tx, creation_bytecode, gas_used)
    VALUES ($1, $2, $3, $4, COALESCE(contract_gas_used.total_gas, 0))
    FROM contract_gas_used
	ON CONFLICT (runtime, contract_address) DO UPDATE
	SET gas_used = gas_used + $5
```

Reply:
It looks like we'll need to be extra careful to make all running total updates commutative from now on; maybe we can add a comment somewhere?

Reply:
Huh good point. That goes back to the discussion on Slack, where you pointed out that dead reckoning for {chain,analysis}.{evm_tokens,evm_contracts} tables is order-sensitive wrt where in the db batch it happens. And before that, when you introduced dead reckoning for total_supply in evm_tokens, the solution with duplicating info across `chain` and `analysis` schemas felt quite hacky (but necessary).

Summarizing from Slack, ligthly edited:
> In my mind, the original source of all of that trouble is that updates to contract addresses (taking contracts as an example; might apply to other tables too) can only happen after we confirm that an address is indeed a contract. Hence the sharing of data between `chain` and `analysis` tables, and complex updating rules.
> So we should try to make those two types of updates ("X is [not] a contract" and "X has used another Y of gas") independent. I suggest we make the table of contracts contain all _potential_ contracts, with an extra bool flag to say if this is really a contract. (Or we can (ab)use the nullness of the bytecode field for that purpose.) The `evm_tokens` table works that already, IIRC; `token_type == 0` means that this contract is not really a token in a meaningful way. If we track contracts this way too, then every dead-reckoning step can be an upsert, rather than an update.

If we do that, the gnarly initial INSERT would go away too, because as soon as we needed a row to increase the `gas_used` of an address, we'd simply create it in `chain.evm_contracts`.

I suggest we merge this PR as-is, and I follow up with a PR that refactors `evm_contracts` as suggested above so that its dead-reckoning updates really are commutative. Same for `evm_tokens`; while we do have copying of dead-reckoned values from `analysis` set up in that table once a candidate-token is confirmed to be a real token, it would be simpler if we immediately inserted into `chain.evm_tokens`.

Reply:
Sounds good!